src/System.Memory/src/System/Buffers/ReadOnlySequence_helpers.cs

```diff
@@ -11,207 +12,171 @@ namespace System.Buffers
+                }
+            }
+            else if (type != SequenceType.Empty && startSegment != endSegment)
+            {
+                ThrowHelper.ThrowInvalidOperationException_EndPositionNotReached();
+            }
 
             switch (type)
             {
                 case SequenceType.MemoryList:
-                    var segment = (IMemoryList<T>)start.Segment;
-                    Memory<T> bufferSegmentMemory = segment.Memory;
-                    int currentEndIndex = bufferSegmentMemory.Length;
-
-                    if (segment == end.Segment)
-                    {
-                        currentEndIndex = endIndex;
-                        next = default;
-                    }
-                    else
-                    {
-                        IMemoryList<T> nextSegment = segment.Next;
-                        if (nextSegment == null)
-                        {
-                            if (end.Segment != null)
-                            {
-                                ThrowHelper.ThrowInvalidOperationException_EndPositionNotReached();
-                            }
-
-                            next = default;
-                        }
-                        else
-                        {
-                            next = new SequencePosition(nextSegment, 0);
-                        }
-                    }
-
-                    data = bufferSegmentMemory.Slice(startIndex, currentEndIndex - startIndex);
-                    return true;
-
-                case SequenceType.OwnedMemory:
-                    var ownedMemory = (OwnedMemory<T>)start.Segment;
-                    if (ownedMemory != end.Segment)
-                    {
-                        ThrowHelper.ThrowInvalidOperationException_EndPositionNotReached();
-                    }
-
-                    data = ownedMemory.Memory.Slice(startIndex, endIndex - startIndex);
-                    next = default;
-                    return true;
-
+                    data = startMemoryList.Memory.Slice(startIndex, length);
+                    break;
                 case SequenceType.Array:
-                    var array = (T[])start.Segment;
-
-                    if (array != end.Segment)
-                    {
-                        ThrowHelper.ThrowInvalidOperationException_EndPositionNotReached();
-                    }
-
-                    data = new Memory<T>(array, startIndex, endIndex - startIndex);
-                    next = default;
-                    return true;
-
+                    data = new ReadOnlyMemory<T>((T[])startSegment, startIndex, length);
+                    break;
+                case SequenceType.OwnedMemory:
+                    data = ((OwnedMemory<T>)startSegment).Memory.Slice(startIndex, length);
+                    break;
+                case SequenceType.String:
+                    data = (ReadOnlyMemory<T>)(object)((string)startSegment).AsMemory(startIndex, length);
```

Comment:
This cast...

Reply:
Standard Jit generics special; else the C# compiler complains. Could otherwise use Unsafe.As

Reply:
What's the resulting asm, is all this optimized out?

Reply:
Yeah, or the boxing in [non-hardware accel Vectors](https://github.com/dotnet/coreclr/blob/35a31158d8a69d978ff61f3b7baf14b505ec5a60/src/mscorlib/shared/System/Numerics/Vector.cs#L268-L365) would be terribad. Don't know what it does for non-struct types though